File: fsharp/queen-attack/QueenAttack.fs
module QueenAttack

let create (position: int * int) =
    match (fst position, snd position) with
    | (x, y) when x >= 0 && x < 8 && y >= 0 && y < 8 -> true
    | _ -> false

let canAttack (queen1: int * int) (queen2: int * int) =
    match (fst queen1, snd queen1, fst queen2, snd queen2) with
    | (x1, _, x2, _) when x1 = x2 -> true
    | (_, y1, _, y2) when y1 = y2 -> true
    | (x1, y1, x2, y2) when abs (x1 - x2) = abs (y1 - y2) -> true
    | _ -> false

File: csharp/queen-attack/QueenAttack.cs
public class Queen
{
    public Queen(int row, int column)
    {
        if (row < 0 || row > 7 || column < 0 || column > 7) throw new ArgumentOutOfRangeException();
        Row = row;
        Column = column;
    }

    public int Row { get; }
    public int Column { get; }
}

public static class QueenAttack
{
    public static bool CanAttack(Queen white, Queen black) => (white, black) switch
    {
        (Queen w, Queen b) when w.Row == b.Row || w.Column == b.Column => true,
        (Queen w, Queen b) when Math.Abs(w.Row - b.Row) == Math.Abs(w.Column - b.Column) => true,
        _ => false
    };

    public static Queen Create(int row, int column) => new(row, column);
}